Improve `InlineDispatcherSwitch` logic

andrealinux1 opened the pull request:
We introduce the following improvements for the `InlineDispatcherSwitch` routine:
- We do not inline `case`s which contain a `break` or `continue` statements. Moving such kind of statement would indeed break the semantics.
- Simplify redundant _non local control flow_ statements: i.e., when we inline a `return` statement, we remove redundant following `break`s and `continue`s.
- We remove the associated `SetNode`s when simplifying `SwitchBreak` statements for exit dispatchers.

Additionally, we remove some dead code and reorder the `SwitchBreaksFixer` pass.

Commit:
BeautifyGHAST: remove unused `using`

Commit:
InlineDispatcherSwitch: remove non-exercised code

Commit:
InlineDispatcherSwitch: do not inline non local CF

Do not inline loop related `break` and `continue` statements.
Indeed, inlining them would mean moving from the scope of a cycle, to an
inner one, non-local control flow statements, and this would break the
semantics.

Commit:
RemoveDeadCode: dead code removal beautify

We introduce a "dead code" optimization like pass on the AST.

The goal is to simplify away, in a `SequenceNode`, everything that
follows a node with sports a `nofallthrough behavior.

Commit:
BeautifyGHAST: reorder `SwitchBreaksFixer` pass

Delay the `SwitchBreaksFixer` beautify pass run.
Specifically, we run it last, as this pass computes information used
only by the backend, and the information it depends upon may be
invalidated by some passes that transform `switch`es (as the
`simplifyDualSwitch` pass).

Commit:
InlineDispatcherSwitch: remove `SetNode`s

During the `SwitchBreakSimplify` pass, we also remove the `SetNode`s
associated to `case` simplified.

pfez commented on include/revng-c/RestructureCFG/ASTNode.h:
```diff
@@ -405,6 +405,8 @@ class SequenceNode : public ASTNode {
 
   links_container::size_type length() const { return NodeVec.size(); }
 
+  bool isEmpty() const { return length() == 0; }
```
Just `empty()` maybe? It's more consistent with STL practice.

andrealinux1 replied:
Ohh right, addressed this in 5cc1ef4720c1a3af2055924147c447000280205f

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
+    SequenceNode *Seq = llvm::cast<SequenceNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (ASTNode *N : Seq->nodes()) {
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(N);
+    }
```
Wouldn't it be faster to iterate in reverse order?
The very moment `HasNonLocalcCF` is set to true, you can just bail out.

andrealinux1 replied:
Right, changed this [here](https://github.com/revng/revng-c/blob/f15b7b1a9506a37ed6d348f7dd6d880b1d93df16/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L519)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
+  } break;
+  case ASTNode::NK_Scs: {
+
+    // We do not want to move around a `ScsNode` for any reason
```
We don't because we already know it creates problems? Or it's just an advanced scenario we haven't fully considered yet?

andrealinux1 replied:
There was actually a bug here, the inlining was always enabled, I now disabled it [here](https://github.com/revng/revng-c/blob/f15b7b1a9506a37ed6d348f7dd6d880b1d93df16/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L532), and added a TODO mentioning that we may want to extend the inlining criterion to cover also this scenario.

The main point is that we could consider this advanced scenario at the cost of moving also the eventual exit dispatcher associated to a cycle when moving the cycle itself. If we do that, I think that we may also enable the inlining for `ScsNode`s.

pfez replied:
> The main point is that we could consider this advanced scenario at the cost of moving also the eventual exit dispatcher associated to a cycle when moving the cycle itself. If we do that, I think that we may also enable the inlining for ScsNodes.

Ok then please expand the comments explaining this. Otherwise, like it is now, it seems that you never want to do it, not that it's a trade-off decision.

andrealinux1 replied:
I motivated this choice, and highlighted that we may want to reverse it, [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L540)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
+    IfNode *If = llvm::cast<IfNode>(Node);
+
+    bool HasNonLocalCF = false;
+    if (If->hasThen()) {
+      ASTNode *Then = If->getThen();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Then);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Else);
+    }
```
Change this code so that it early returns `true` when possible.

Like, if the `then` has non-local control flow, there's no need to look inside the `else`.

andrealinux1 replied:
Right, applied the suggested changes [here](https://github.com/revng/revng-c/blob/f15b7b1a9506a37ed6d348f7dd6d880b1d93df16/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L539)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (auto &LabelCasePair : Switch->cases()) {
+      HasNonLocalCF = HasNonLocalCF
+                      or rc_recur containsNonLocalCF(LabelCasePair.second);
+    }
```
Same as above, you can early exit at the first case that has non-local cf.

andrealinux1 replied:
Implemented the optimization [here](https://github.com/revng/revng-c/blob/f15b7b1a9506a37ed6d348f7dd6d880b1d93df16/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L560)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
+    IfNode *If = llvm::cast<IfNode>(Node);
+
+    bool HasNonLocalCF = false;
+    if (If->hasThen()) {
+      ASTNode *Then = If->getThen();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Then);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Else);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (auto &LabelCasePair : Switch->cases()) {
+      HasNonLocalCF = HasNonLocalCF
+                      or rc_recur containsNonLocalCF(LabelCasePair.second);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Set:
+  case ASTNode::NK_Code:
+  case ASTNode::NK_SwitchBreak: {
+    rc_return false;
+  } break;
```
I'm not 100% sure about the `SwitchBreak`. Can you please elaborate?

andrealinux1 replied:
The reasoning is the following:
- A `SwitchBreak` can appear in the AST as the single element composing the body of a `case` of a `switch` node. Its meaning is equivalent to `nop` (e.g., a certain state variable assignment of an exit dispatcher corresponds to no statements executed).
- During the `inlineDispatcherSwitch` pass, in principle we can remove a `case` composed by a single `SwitchBreak` node. However, in the current state of things, this is done before in a dedicated pass called `simplifySwitchBreak`, indeed there is a specific point where we check that we do not encounter any `SwitchBreak` in this phase ([this check here](https://github.com/andrealinux1/revng-c/blob/954765bd27fbbbd5b28e738c30591490f8dc315e/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L707)).
- However, I share your concern that this is not crystal clear to understand.
- At the same time, I think that the semantics of the `containsNonLocalCF` helper function should handle the presence of a `SwitchBreak`. Maybe the non ideal thing here is the name of the helper itself. Basically we want an helper which replies "no you can't move around this `ASTNode`" only when it finds non local control flow related to loops (standard `break`s and `continue`s)

pfez replied:
I understand now, and I agree with your last statement.

Can you propose better naming for the helper?

andrealinux1 replied:
I changed the name of the helper [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L521), in `containsContinueOrBreak`

pfez replied:
Much better, thanks!

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
   RootNode = inlineDispatcherSwitchImpl(AST, RootNode, LoopDispatcherMap);
 
+  // Pre-compute the `FallThroughScopeType` before the `SuperfluousNonLocalCF`
+  // simplidy is performed. This information is useful in order to decide
```
Typo: simpliDy

andrealinux1 replied:
With the introduction of the new `RemoveDeadCode` beautify pass this portion of code want away.

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+  return RootNode;
+}
+
+ASTNode *inlineDispatcherSwitch(const model::Binary &Model,
+                                ASTTree &AST,
+                                ASTNode *RootNode) {
```
At this point I don't think you even need to take the `RootNode` argument explicitly.
It seems to me that the whole code always assumes that `RootNode == AST.getRoot()`. Isn't that the case?
If I'm right, passing `RootNode` explicitly opens up opportunities for callers to pass a wrong `RootNode`, so I'd just remove it.

Same holds for `inlineDispatcherSwitch` below. Please fix that too.

andrealinux1 replied:
I crafted a new commit dedicated to removing all the unnecessary `RootNode` argument b9a8a08f7f6160f57da4506aea7fce6f85550453

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+        // processed
+        N = rc_recur simplifySuperfluousNonLocalCFImpl(N, FallThroughScopeMap);
+        FallThroughScopeType ReturnType = FallThroughScopeType::Return;
+        ShouldSimplify = (FallThroughScopeMap.at(N) == ReturnType);
```
Why the parentheses on RHS? Please drop them.

andrealinux1 replied:
In the new version of the code, the parentheses are actually needed [here](https://github.com/revng/revng-c/blob/266c681abb4668354bbbbc5cb6802ad4c583ac30/lib/RestructureCFG/RemoveDeadCode.cpp#L37)

pfez replied:
Well use `!=` and drop the parentheses then

andrealinux1 replied:
Right, changed this [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/RemoveDeadCode.cpp#L37)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+      }
+    }
+
+    // In this beautify, it may be that a dispatcher switch is completely
+    // removed, therefore leaving a `nullptr` in a `SequenceNode`. We remove all
+    // these `nullptr` from the `SequenceNode` after the processing has taken
+    // place
+    Seq->removeNode(nullptr);
```
Is this guaranteed to always remove all of the `nullptr`s left? Or only the first one?

andrealinux1 replied:
Yes, the `removeNode` method of  `SequenceNode` takes care of removing all the instances of empty nodes

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+    // place
+    Seq->removeNode(nullptr);
+
+    // This simplify should never completely empty a `SequenceNode`
+    revng_assert(not Seq->isEmpty());
```
So, worst case, the sequence is left with a single node.

Don't we have an invariant that requires sequence composed of only 1 node to be unwrapped?

andrealinux1 replied:
You are right, but there are some caveats that we informally discussed in the past, briefly:
- In principle, we could run the three main steps to enforce all the invariants at each beautify step, by invoking the `normalize` method defined [here](https://github.com/revng/revng-c/blob/9624f2bbd14b6a3aead30f7c96fc6ed9a0b255ad/include/revng-c/RestructureCFG/GenerateAst.h#L1005), but this would mean traversing the AST multiple times.
- Therefore, we decided in the past to keep the passes that enforce a single invariant separately available.
-  And this is the case we are in. If you take a look a the beautify passes "pipeline" [here](https://github.com/revng/revng-c/blob/9624f2bbd14b6a3aead30f7c96fc6ed9a0b255ad/lib/RestructureCFG/BeautifyGHAST.cpp#L1025), you can see that the `simplifyAtomicSequence` pass (which is in charge to unwrap unitary sequences), is run after `inlineDispatcherSwitch`, therefore eventually enforcing the property you mentioned, with the assurance that with the current anatomy of the "pipeline" we do not have any other pass that requires such property enforced, between `inlineDispatcherSwitch` and `simplifyAtomicSequence` itself.

- We could, o.c., run it just after `inlineDispatcherSwitch` , and this would prevent possible problems with future reordering of the steps of the "pipeline", at the cost of visiting more times the AST. I think it could be worth discussing about this IRL if you want more details.

pfez replied:
Mmmh ok it makes sense like it is. Just note it in the comments.

andrealinux1 replied:
I added some comments commenting on the motivation for this assertion [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/RemoveDeadCode.cpp#L49)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -968,12 +980,44 @@ static RecursiveCoroutine<ASTNode *> simplifySwitchBreakImpl(ASTNode *Node) {
+          for (auto *Set : Sets) {
+
+            // To perform the removal, we reuse the `addToDispatcherSet`
+            addToDispatcherSet(AST, RelatedLoop->getBody(), nullptr, Set, true);
```
Please make a dedicated directive for this. Using `addToDispatcherSet` with `nullptr` leaks internal details and is slightly confusing if you ignore those details.

andrealinux1 replied:
Right, I created a new wrapper to hide the internal working details [here](https://github.com/revng/revng-c/blob/4c08f81ad95989ec412c54a03f1e80560eef7a0a/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L226)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -1002,8 +1046,11 @@ static RecursiveCoroutine<ASTNode *> simplifySwitchBreakImpl(ASTNode *Node) {
 /// `case`.
 ASTNode *simplifySwitchBreak(ASTTree &AST, ASTNode *RootNode) {
```
Here too I think we don't need the `RootNode` argument. All the code assumes `RootNode` is the root of `AST`, so `AST` is enough.

andrealinux1 replied:
Done this in a dedicated commit b9a8a08f7f6160f57da4506aea7fce6f85550453

pfez replied:
:+1:

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -960,6 +969,9 @@ static RecursiveCoroutine<ASTNode *> simplifySwitchBreakImpl(ASTNode *Node) {
       llvm::SmallVector<size_t> ToRemoveCaseIndex;
 
+      // Save the `LabelSet`s associated to the removed `case`s
+      llvm::SmallSet<size_t, 1> RemovedAtomicLabels;
```
Why `Atomic`?

pfez replied:
Does that just mean that they have a single label and not many? If so, drop `Atomic` from the name, it seems to hint at some atomicity of the execution or something, which isn't the case.

andrealinux1 replied:
The `Atomic` meaning was the one you highlighted. Removed it [here](https://github.com/revng/revng-c/blob/4c08f81ad95989ec412c54a03f1e80560eef7a0a/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L870)

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
 
+  // Pre-compute the `FallThroughScopeType` before the `SuperfluousNonLocalCF`
+  // simplidy is performed. This information is useful in order to decide
+  // whether we remove some statements that after the `case` inlining are
+  // preceded by `return` statements.
+  FallThroughScopeTypeMap
+    FallThroughScopeMap = computeFallThroughScope(Model, RootNode);
+
+  // Perform the `SuperfluousNonLocalCF` simplification pass
+  RootNode = simplifySuperfluousNonLocalCF(AST, RootNode, FallThroughScopeMap);
+
```
This looks like a dead-code-elimination pass.
Please carve this out in a separate beautifier.
And give it a name like `removeDeadCode`.

andrealinux1 replied:
I implemented a `removeDeadCode` beautify pass in a separate commit 266c681abb4668354bbbbc5cb6802ad4c583ac30, so everything is more autocontained

pfez commented on lib/RestructureCFG/InlineDispatcherSwitch.cpp:
```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+        // Compute the `FallThrough information after each `SequencenNode` is
+        // processed
+        N = rc_recur simplifySuperfluousNonLocalCFImpl(N, FallThroughScopeMap);
+        FallThroughScopeType ReturnType = FallThroughScopeType::Return;
```
We should do something about calls to noreturn functions too.

andrealinux1 replied:
In rewriting this as the `RemoveDeadCode` beautify pass, I changed the approach here. Basically, I know look for everything that does not `fallthrough`, and this encompasses `noreturn` functions too. Code [here](https://github.com/revng/revng-c/blob/266c681abb4668354bbbbc5cb6802ad4c583ac30/lib/RestructureCFG/RemoveDeadCode.cpp#L36)

Commit:
FallThroughScope: `ScsNode`s do fallthrough

The `FallThroughScope` associated to a `ScsNode` should represent the
fact that we have fallthrough from a loop.

Without any more advanced semantical analysis, we cannot conclude that
we do not have fallthrough.

Commit:
FallThroughScope: dispatcher `switch` fallthroughs

Change the analysis in order to lift the assumption that a dispatcher
`switch` always coveris all the values of the variable state. This is
not true after the `simplifySwitchBreak` and `inlineDispatcherSwitch`
beautify passes, which can remove some of the `case`s.

Commit:
FallThroughScope: remove `SwitchBreak` type

A `SwitchBreak` node should not have a specific associated
`FallThroughScopeType`, since its semantics represent the fact of a
`case` of the `switch` with no associated statements, and thus can be
represented with the `FallThrough` behavior.

Commit:
ASTNode: STL practice consistency

Commit:
InlineDispatcherSwitch: improve function signature

andrealinux1 commented:
@pfez One additional thing that I want to mention is that with the introduction of the `removeDeadCode` beautify pass, I found, investigated and squashed a couple of additional bugs that were caused by incorrect semantics of the routine that computes the `FallThroughScope` property of each AST node.

You can take a look at these changes in the following commits: 09204508f87acfcf025f818db4c80fbaf9970893, f6bedcf04ea66e64b1b02a4d604c74195eb1f235, 32fdc5b8072648ec0686b2f5017423e5a449cf05.

I think that another couple of eyes are useful on this, since this pass can prune away parts of the AST. I also have a couple of more elaborated observations on this, maybe ping me IRL when reviewing this and we can discuss them.

aleclearmind commented:
Never call branches `fix/*`. We only have `feature/*` branches.

andrealinux1 commented:
> Never call branches `fix/*`. We only have `feature/*` branches.

@aleclearmind ACK, shall I rename the current branch for this PR?

pfez commented on include/revng-c/RestructureCFG/ASTNode.h:
```diff
@@ -405,6 +405,8 @@ class SequenceNode : public ASTNode {
 
   links_container::size_type length() const { return NodeVec.size(); }
 
+  bool isEmpty() const { return NodeVec.empty(); }
```
If this is really for consistency with STL then `s/isEmpty/empty` please.

andrealinux1 replied:
Changed this [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/include/revng-c/RestructureCFG/ASTNode.h#L408)

pfez commented on lib/RestructureCFG/BeautifyGHAST.cpp:
```diff
@@ -1024,6 +1025,11 @@ void beautifyAST(const model::Binary &Model, Function &F, ASTTree &CombedAST) {
   RootNode = inlineDispatcherSwitch(CombedAST);
   Dumper.log("after-dispatcher-switch-inlining");
 
+  // Perform the dead code simplification
```
Please add more explanation about why it makes sense to call this here and not any other places.

andrealinux1 replied:
Added some related comments [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/BeautifyGHAST.cpp#L1028)
